aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -214,40 +214,45 @@ void TensorIterator::allocate_outputs() {
 }
 
 #ifdef BUILD_NAMEDTENSOR
-void TensorIterator::propagate_names_to_outputs() {
-  bool should_perform_name_inference = std::any_of(
+void TensorIterator::compute_names() {
+  bool should_infer_names = std::any_of(
       operands_.begin(),
       operands_.end(),
       [](const OperandInfo& op) {
         return op.tensor.defined() && op.tensor.has_names();
       });
-  if (!should_perform_name_inference) {
-    return;
-  }
 
-  // build names
-  NameVector names;
   for (auto& op : operands_) {
     if (!op.tensor.defined()) continue;
     // don't include output tensors that are not also input tensors.
     if (resize_outputs_ && op.is_output && !op.is_read_write) continue;
     // perform name inference
-    if (names.empty()) {
-      names = op.tensor.names();
+    if (names_.empty()) {
+      names_ = op.tensor.names();
     } else {
-      names = NameVector(unify_from_right(names, op.tensor.names()));
+      names_ = NameVector(unify_from_right(names_, op.tensor.names()));
     }
   }
+}
+
+void TensorIterator::propagate_names_to_outputs() {
+  // names_ can be empty for two reasons:
+  // 1. We were performing ops on scalar tensors. Then there should be no names.
+  // 2. All of the defined inputs/outputs had no names. Then we shouldn't
+  //    run name inference.
+  if (names_.empty()) {
+    return;
+  }
 
   // propagate names
   for (int i = 0; i < num_outputs_; i++) {
```

Comment:
nit: not really related to your change but I might prefer `i < noutputs()` and `outputs(i)` here.

Reply:
Glancing through the TensorIterator code, I see that `num_outputs_` is used a lot instead of `noutputs()`. I'm not sure what the concerns are, maybe `noutputs()` is more of the outward-facing API.

`outputs(i)` performs an unnecessary `AT_ASSERT`. We don't compile those away and until we do, putting it in a loop like this probably won't let us squeeze out the last bit of performance that we're trying to get.

Reply:
I just prefer it because `operands_[i]` may not be clear to some readers since it relies on knowing that outputs are stored at the beginning of the operands array. I'm not sure what the perf overhead is and we do use `operands_[i]` elsewhere here, so not a review-blocker, just a preference of mine.